crawl.go

```diff
@@ -0,0 +1,198 @@
+	"/dns4/bootstrap-3.calibration.fildev.network/tcp/1347/p2p/12D3KooWMHDi3LVTFG8Szqogt7RkNXvonbQYqSazxBx41A5aeuVz",
+}
+
+type FilecoinPeer struct {
+	ctx            context.Context
+	dht            *dht.IpfsDHT
+	host           host.Host
+	bootstrapPeers []peer.AddrInfo
+}
+
+func FilecoinBootstrapPeers(peerList []string) []peer.AddrInfo {
+	if peerList == nil {
+		peerList = MainnetPeers
+	}
+
+	maddrs := make([]multiaddr.Multiaddr, len(peerList))
+
+	for i, bp := range peerList {
+		var err error
+		maddrs[i], err = multiaddr.NewMultiaddr(bp)
+		if err != nil {
+			return nil
+		}
+	}
+
+	peers, _ := peer.AddrInfosFromP2pAddrs(maddrs...)
+
+	return peers
+}
+
+func SetupFilecoinPeer(
+	ctx context.Context,
+	network NetworkName,
+) (*FilecoinPeer, error) {
+	if network == "" {
+		network = Mainnet
+	}
+
+	var pp protocol.ID
+	var bootstrapPeers []peer.AddrInfo
+
+	switch network {
+	case Mainnet:
+		pp = "/fil/kad/testnetnet"
+		bootstrapPeers = FilecoinBootstrapPeers(MainnetPeers)
+	case Calibnet:
+		pp = "/fil/kad/calibnet"
+		bootstrapPeers = FilecoinBootstrapPeers(CalibnetPeers)
+	}
+
+	var ddht *dht.IpfsDHT
+	priv, _, err := crypto.GenerateKeyPair(crypto.RSA, 2048)
+
+	if err != nil {
+		return &FilecoinPeer{}, err
+	}
+
+	libp2pOpts := []libp2p.Option{
+		libp2p.ListenAddrStrings(
+			"/ip4/0.0.0.0/tcp/9000",      // regular tcp connections
+			"/ip4/0.0.0.0/udp/9000/quic", // a UDP endpoint for the QUIC transport
+		),
+		libp2p.Identity(priv),
+		libp2p.Security(libp2ptls.ID, libp2ptls.New),
+		libp2p.Security(noise.ID, noise.New),
+		libp2p.DefaultTransports,
+		libp2p.Routing(func(h host.Host) (routing.PeerRouting, error) {
+			ddht, err := dht.New(
+				ctx,
+				h,
+				dht.Option(dht.Datastore(dsync.MutexWrap(ds.NewMapDatastore()))),
+				dht.Option(dht.NamespacedValidator("pk", record.PublicKeyValidator{})),
+				dht.Option(dht.NamespacedValidator("ipns", ipns.Validator{KeyBook: h.Peerstore()})),
+				dht.Option(dht.Concurrency(10)),
+				dht.Option(dht.Mode(dht.ModeAuto)),
+				dht.Option(dht.ProtocolPrefix(pp)),
+			)
+
+			return ddht, err
+		}),
+		libp2p.NATPortMap(),
+		libp2p.ConnectionManager(connmgr.NewConnManager(100, 600, time.Minute)),
+		libp2p.EnableAutoRelay(),
+		libp2p.EnableNATService(),
+	}
+
+	h, err := libp2p.New(libp2pOpts...)
+
+	if err != nil {
+		return &FilecoinPeer{}, err
+	}
+
+	return &FilecoinPeer{ctx: ctx, host: h, dht: ddht, bootstrapPeers: bootstrapPeers}, nil
+}
+
+func (f *FilecoinPeer) Bootstrap() {
+	connected := make(chan struct{})
+
+	var wg sync.WaitGroup
+	for _, pinfo := range f.bootstrapPeers {
+		wg.Add(1)
+		go func(p peer.AddrInfo) {
+			defer wg.Done()
+			err := f.host.Connect(f.ctx, p)
+			if err != nil {
+				log.Println(err)
+				return
+			}
+			log.Println("Connected to ", p.ID)
+			connected <- struct{}{}
+		}(pinfo)
+	}
+
+	go func() {
+		wg.Wait()
+		close(connected)
+	}()
+}
+
+func (f *FilecoinPeer) Crawl(d time.Duration) {
```

Comment:
I find this is not really doing any Crawling, just bootstrapping and logging stats. I suggest this is left only to print stats, and bootstrap is called separately from somewhere else (i.e. after creation of the peer).